lib/src/subscription.dart

```diff
@@ -0,0 +1,155 @@
+    if (other is! Subscription) return false;
+    return realmCore.subscriptionEquals(this, other);
+  }
+}
+
+extension SubscriptionInternal on Subscription {
+  SubscriptionHandle get handle => _handle;
+}
+
+class _SubscriptionIterator implements Iterator<Subscription> {
+  int _index = -1;
+  final SubscriptionSet _subscriptions;
+
+  _SubscriptionIterator._(this._subscriptions);
+
+  @override
+  Subscription get current => _subscriptions.elementAt(_index);
+
+  @override
+  bool moveNext() => ++_index < _subscriptions.length;
+}
+
+enum SubscriptionSetState {
+  uncommitted,
+  pending,
+  bootstrapping,
+  complete,
+  error,
+  superseded,
+}
+
+abstract class SubscriptionSet with IterableMixin<Subscription> {
+  Realm _realm;
+  SubscriptionSetHandle _handle;
+
+  SubscriptionSet._(this._realm, this._handle);
+
+  Subscription? find<T extends RealmObject>(RealmResults<T> query) {
+    return realmCore.findSubscriptionByQuery(this, query).convert(Subscription._);
+  }
+
+  Subscription? findByName(String name) {
+    return realmCore.findSubscriptionByName(this, name).convert(Subscription._);
+  }
+
+  Future<SubscriptionSetState> waitForStateChange(SubscriptionSetState state) async {
+    return SubscriptionSetState.values[await realmCore.waitForSubscriptionSetStateChange(this, state)];
+  }
+
+  @override
+  int get length => realmCore.getSubscriptionSetSize(this);
+
+  @override
+  Subscription elementAt(int index) {
+    return Subscription._(realmCore.subscriptionAt(this, index));
+  }
+
+  Subscription operator [](int index) => elementAt(index);
+
+  @override
+  _SubscriptionIterator get iterator => _SubscriptionIterator._(this);
+
+  void update(void Function(MutableSubscriptionSet mutableSubscriptions) action);
+
+  int get version => realmCore.subscriptionSetVersion(this);
+
+  SubscriptionSetState get state => SubscriptionSetState.values[realmCore.subscriptionSetState(this)];
+}
+
+extension SubscriptionSetInternal on SubscriptionSet {
+  Realm get realm => _realm;
+  SubscriptionSetHandle get handle => _handle;
+
+  static SubscriptionSet create(Realm realm, SubscriptionSetHandle handle) => MutableSubscriptionSet._(realm, handle);
+}
+
+class MutableSubscriptionSet extends SubscriptionSet {
+  MutableSubscriptionSetHandle? _mutableHandle;
+
+  MutableSubscriptionSet._(Realm realm, SubscriptionSetHandle handle) : super._(realm, handle);
+
+  @override
+  void update(void Function(MutableSubscriptionSet mutableSubscriptions) action) {
+    assert(_mutableHandle == null);
+    var commit = false;
+    try {
+      _mutableHandle = realmCore.makeSubscriptionSetMutable(this);
+      action(this);
+      commit = true;
+    } finally {
+      if (commit) {
+        _handle = realmCore.subscriptionSetCommit(this);
+      }
+      // _mutableHandle.release(); // TODO: Release early
+      _mutableHandle = null;
+    }
+  }
+
+  bool addOrUpdate<T extends RealmObject>(RealmResults<T> query, {String? name}) {
```

Comment:
Also, it's a good idea to return the newly added subscription from add.

Reply:
The C-API only exposes an `addOrUpdate` kind of method, aka `realm_sync_subscription_set_insert_or_assign`. We return `true` if it was actually added (/inserted). 

We could return the newly added subscription, but users might do something stupid like storing it in a variable, before it is commit'ed. Isn't it safer that they will need to wait until after the commit to actually retrieve it? Otherwise we need to handle managed vs. unmanaged subscriptions.

Reply:
Not sure I understand the distinction there - when you add a subscription to the set, that subscription is already managed.

Reply:
.. but not commit'ed.. it may never be

Reply:
I'll make `addOrUpdate` return a `Subscription`. I will not add an `update` flag since you cannot set it to false.

Reply:
While you can't set it to `true`, you can implement the logic to check for duplicates in the SDK. I believe all SDKs offer ability to set an `update` flag and we should strive for consistency there.

Reply:
I relent. I will fix this.